gadgets/mri_core/ImageArraySplitGadget.cpp

```diff
@@ -33,60 +19,50 @@ int ImageArraySplitGadget::process( GadgetContainerMessage<IsmrmrdImageArray>* m
-                GadgetContainerMessage< hoNDArray< std::complex<float> > >* cm2 = 
-                        new GadgetContainerMessage<hoNDArray< std::complex<float> > >();
+    // Loop over N, S and LOC
+    for (uint16_t loc = 0; loc < LOC; loc++) {
+        for (uint16_t s = 0; s < S; s++) {
+            for (uint16_t n = 0; n < N; n++) {
+                // Create a new image header and copy the header for this n, s and loc
+                auto imageHeader = ISMRMRD::ImageHeader();
+                memcpy(&imageHeader, &imagearr.headers_(n, s, loc), sizeof(ISMRMRD::ImageHeader));
 
-                try{cm2->getObjectPtr()->create(img_dims);}
-                catch (std::runtime_error &err){
-                    GEXCEPTION(err,"Unable to allocate new image\n");
-                    cm1->release();
-                    cm2->release();
-                    return GADGET_FAIL;
-                }
-                memcpy(cm2->getObjectPtr()->get_data_ptr(), &imagearr.data_(0,0,0,0,n,s,loc), X*Y*Z*CHA*sizeof(std::complex<float>));
-                //Chain them
-                cm1->cont(cm2);
+                // Create a new image image and the 4D data block [X,Y,Z,CHA] for this n, s and loc
+                auto imageData = hoNDArray<std::complex<float>>(img_dims);
+                memcpy(imageData.get_data_ptr(), &imagearr.data_(0, 0, 0, 0, n, s, loc), X * Y * Z * CHA * sizeof(std::complex<float>));
 
-                //Create a new meta container if needed and copy
-                if (imagearr.meta_.size()>0) {
-                    GadgetContainerMessage< ISMRMRD::MetaContainer >* cm3 = 
-                            new GadgetContainerMessage< ISMRMRD::MetaContainer >();
-                    size_t mindex = loc*N*S + s*N + n;
-                    *cm3->getObjectPtr() = imagearr.meta_[mindex];
-                    cm2->cont(cm3);
+                // Create a new meta container if needed and copy
+                auto imageMetaContainer = std::optional<ISMRMRD::MetaContainer>();
+                if (imagearr.meta_.size() > 0) {
+                    size_t mindex = loc * N * S + s * N + n;
+                    imageMetaContainer = imagearr.meta_[mindex];
                 }
 
-                //Pass the image down the chain
-                if (this->next()->putq(cm1) < 0) {
-                    return GADGET_FAIL;
-                }
+                // Pass the image down the chain
+                out.push(Core::Image<std::complex<float>>(std::move(imageHeader), std::move(imageData), std::move(imageMetaContainer)));
             }
         }
     }
-    
-    m1->release();
-    return GADGET_OK;  
-
 }
+} // namespace
 
-GADGET_FACTORY_DECLARE(ImageArraySplitGadget)
+namespace Gadgetron {
+
+void ImageArraySplitGadget::process(Core::InputChannel<ImageOrImageArray>& in, Core::OutputChannel& out) {
+    auto lambda = [&](auto message){ 
+        splitInputData(message, out);
+    };
+    for (auto msg : in) {
+        visit(lambda, msg);
```

Comment:
I would inline the lambda here, but that is more of a style choice.

Reply:
Agreed - another remnant from me learning how all the syntax was working. Fixed.